src/controller/CHIPDeviceController.cpp

```diff
@@ -231,47 +231,90 @@ CHIP_ERROR DeviceController::Init(NodeId localDeviceId, ControllerInitParams par
+        PeerId().SetNodeId(deviceId), CSR, 1, noc.Get(), kMaxCHIPOpCertLength, nocLen));
 
-    if (!admin->AreCredentialsAvailable())
+    ReturnErrorCodeIf(nocLen == 0, CHIP_ERROR_CERT_NOT_FOUND);
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> ica;
+    ReturnErrorCodeIf(!ica.Alloc(kMaxCHIPOpCertLength), CHIP_ERROR_NO_MEMORY);
+
+    uint32_t icaLen = 0;
+    ChipLogProgress(Controller, "Getting intermediate CA certificate from the issuer");
+    CHIP_ERROR err = mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, ica.Get(), kMaxCHIPOpCertLength, icaLen);
+    ChipLogProgress(Controller, "GetIntermediateCACertificate returned %" PRId32, err);
+    if (err == CHIP_ERROR_INTERMEDIATE_CA_NOT_REQUIRED)
+    {
+        // This implies that the commissioner application uses root CA to sign the operational
+        // certificates, and an intermediate CA is not needed. It's not an error condition, so
+        // let's just send operational certificate and root CA certificate to the device.
+        icaLen = 0;
+        ChipLogProgress(Controller, "Intermediate CA is not needed");
+    }
+    else if (err != CHIP_NO_ERROR)
     {
-        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer1;
-        ReturnErrorCodeIf(!buffer1.Alloc(kMaxCHIPCSRLength), CHIP_ERROR_NO_MEMORY);
+        return err;
+    }
 
-        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer2;
-        ReturnErrorCodeIf(!buffer2.Alloc(kMaxCHIPOpCertLength), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(ConvertX509CertsToChipCertArray(ByteSpan(noc.Get(), nocLen), ByteSpan(ica.Get(), icaLen), certBuf,
+                                                         certBufSize, outCertLen));
 
-        uint8_t * CSR    = buffer1.Get();
-        size_t csrLength = kMaxCHIPCSRLength;
-        ReturnErrorOnFailure(keypair->NewCertificateSigningRequest(CSR, csrLength));
+    return CHIP_NO_ERROR;
+}
 
-        uint8_t * cert   = buffer2.Get();
-        uint32_t certLen = 0;
+CHIP_ERROR DeviceController::LoadLocalCredentials(Transport::AdminPairingInfo * admin)
+{
+    ChipLogProgress(Controller, "Getting operational keys");
+    Crypto::P256Keypair * keypair = admin->GetOperationalKey();
 
-        // TODO - Match the generated cert against CSR and operational keypair
-        //        Make sure it chains back to the trusted root.
-        ChipLogProgress(Controller, "Generating operational certificate for the controller");
-        ReturnErrorOnFailure(mOperationalCredentialsDelegate->GenerateNodeOperationalCertificate(
-            PeerId().SetNodeId(mLocalDeviceId), ByteSpan(CSR, csrLength), 1, cert, kMaxCHIPOpCertLength, certLen));
+    ReturnErrorCodeIf(keypair == nullptr, CHIP_ERROR_NO_MEMORY);
 
-        uint8_t * chipCert   = buffer1.Get();
+    if (!admin->AreCredentialsAvailable())
+    {
+        chip::Platform::ScopedMemoryBuffer<uint8_t> chipCert;
+        uint32_t chipCertAllocatedLen = kMaxCHIPOpCertLength * 2;
```

Comment:
we are asking for 2K of contigous memory here and then another 1K of RAM in the sub-processing below.

Will this work for embedded devices or is spec forcing this on us? Should we maybe  add a note that we may want to allocate these buffers in a separate static-allocated pool to guarantee space availabilty?

Reply:
Added some documentation to it.